sklearn/tree/_criterion.pyx

```diff
@@ -1173,17 +1175,174 @@ cdef class MSE(RegressionCriterion):
+        r = ranks[p]
+        tree.add(r, sorted_y[r], w)
+
+        # Weighted alpha-quantile by cumulative weight
+        half_weight = 0.5 * tree.total_w
+        median_idx = tree.search(half_weight, &w_left, &wy_left, &median_prev_idx)
+
+        if median_idx != median_prev_idx:
+            # Exact match for half_weight in the tree, take the middle point:
+            median = (sorted_y[median_prev_idx] + sorted_y[median_idx]) / 2
+        else:
+            median = sorted_y[median_idx]
+
+        w_right = tree.total_w - w_left
+        wy_right = tree.total_wy - wy_left
+
+        # O(1) pinball loss formula
+        medians[p] = median
+        abs_errors[p] += (
+            (wy_right - median * w_right)
+            + (median * w_left - wy_left)
+        )
+        p += step
+
+
+cdef inline void compute_ranks(
+    float64_t* sorted_y,
+    intp_t* sorted_indices,
+    intp_t* ranks,
+    intp_t n
+) noexcept nogil:
+    """Sort `sorted_y` inplace and fill `ranks` accordingly"""
+    cdef intp_t i
+    for i in range(n):
+        sorted_indices[i] = i
+    sort(sorted_y, sorted_indices, n)
+    for i in range(n):
+        ranks[sorted_indices[i]] = i
+
+
+def _py_precompute_absolute_errors(
+    const float64_t[:, ::1] ys,
+    const float64_t[:] sample_weight,
+    const intp_t[:] sample_indices,
+    const intp_t start,
+    const intp_t end,
+    const intp_t n,
+):
+    """Used for testing precompute_absolute_errors."""
+    cdef:
+        intp_t p, i
+        intp_t s = start
+        intp_t e = end
+        WeightedFenwickTree tree = WeightedFenwickTree(n)
+        float64_t[::1] sorted_y = np.empty(n, dtype=np.float64)
+        intp_t[::1] sorted_indices = np.empty(n, dtype=np.intp)
+        intp_t[::1] ranks = np.empty(n, dtype=np.intp)
+        float64_t[::1] abs_errors = np.zeros(n, dtype=np.float64)
+        float64_t[::1] medians = np.empty(n, dtype=np.float64)
+
+    if start > end:
+        s = end + 1
+        e = start + 1
+    for p in range(s, e):
+        i = sample_indices[p]
+        sorted_y[p - s] = ys[i, 0]
+    compute_ranks(&sorted_y[0], &sorted_indices[0], &ranks[s], n)
+
+    precompute_absolute_errors(
+        sorted_y, ranks, sample_weight, sample_indices, tree,
+        start, end, abs_errors, medians
+    )
+    return np.asarray(abs_errors)[s:e], np.asarray(medians)[s:e]
+
+
+cdef class MAE(Criterion):
     r"""Mean absolute error impurity criterion.
 
-       MAE = (1 / n)*(\sum_i |y_i - f_i|), where y_i is the true
-       value and f_i is the predicted value."""
+    It has almost nothing in common with other regression criterions
+    so it doesn't inherit from RegressionCriterion.
 
-    cdef cnp.ndarray left_child
-    cdef cnp.ndarray right_child
-    cdef void** left_child_ptr
-    cdef void** right_child_ptr
+    MAE = (1 / n)*(\sum_i |y_i - f_i|), where y_i is the true
+    value and f_i is the predicted value.
+    In decision trees, all samples within a node share the same f_i value,
```

Comment:
I think `f_i` is used elsewhere to denote the value of a the i-th feature in `X`. I would rename to `p_i` (p for prediction) or `v_i` (v for leaf "value") to avoid confusion.